File: 1cuatri/ED/final/2023-sept/haskell/SparseMatrix.hs
-------------------------------------------------------------------------------
-- Student's name: ?????????????????????????????????????
-- Identity number (DNI if Spanish/passport if Erasmus): ???????????????????
-- Student's group: ?
-- PC code: ???
--
-- Data Structures. Grado en Informatica. UMA.
-------------------------------------------------------------------------------

module SparseMatrix(SparseMatrix, set, get, add, transpose, fromList) where

import qualified DataStructures.Dictionary.AVLDictionary as D 
import Data.Maybe

data Index = Idx Int Int deriving (Eq, Ord, Show)

-- | = Exercise a - sparseMatrix
data SparseMatrix = SM Int Int (D.Dictionary Index Int) deriving Show 

sparseMatrix :: Int -> Int -> SparseMatrix
sparseMatrix f c = SM f c D.empty -- to complete

-- | = Exercise b - value
value :: SparseMatrix -> Index -> Int
value sm@(SM f1 c1 d) i@(Idx f2 c2)
    | D.isDefinedAt i d     = fromJust(D.valueOf i d)
    | otherwise             = 0

-- | = Exercise c - update
update :: SparseMatrix -> Index -> Int -> SparseMatrix
update sm@(SM f1 c1 d) i@(Idx f2 c2) v = SM f1 c1 (D.insert i v d)

-- | = Exercise d - index
index :: SparseMatrix -> Int -> Int -> Index
index sm@(SM f1 c1 d) f c
    | f >= f1 || f < 0 || c >= c1 || c < 0      = error "index: index is not valid"
    | otherwise                                 = Idx f c

-- | = Exercise e - set
set :: SparseMatrix -> Int -> Int -> Int -> SparseMatrix
set sm f c v = update sm (index sm f c) v

-- | = Exercise f - get
get :: SparseMatrix -> Int -> Int -> Int
get sm f c = value sm (index sm f c)

-- | = Exercise g - add
add :: SparseMatrix -> SparseMatrix -> SparseMatrix
add sm1 sm2@(SM f2 c2 d2) = aux sm1 sm2 (D.keys d2)
    where
        aux sm1 sm2 []          = sm1
        aux sm1 sm2 (k2:ks)     = aux sm1' sm2 (ks)
            where
                sm1' = set sm1 f' c' ((value sm1 k2) + (value sm2 k2))
                (Idx f' c') = k2

-- | = Exercise h - transpose
transpose :: SparseMatrix -> SparseMatrix
transpose sm@(SM f c d) = aux sm (D.keys d)
    where
        aux sm []       = sm
        aux sm (k:ks)   = aux sm'' ks
            where
                (SM f2 c2 d2) = set sm c1 f1 v
                sm'' = (SM f2 c2 (D.delete k d2))

                v = value sm k
                (Idx f1 c1) = k

-- | = Exercise i - fromList
-- Complexity of fromList: O(n*logn) -- to complete
fromList :: Int -> Int -> [Int] -> SparseMatrix
fromList f c list
    | (length list) `mod` 3 == 0    = aux (sparseMatrix f c) list
    | otherwise                     = error "fromList: list is not multiple of 3"
        where
            aux sm []               = sm
            aux sm (f1:c1:v1:xs)    = aux (set sm f1 c1 v1) xs




-- Examples

m1 :: SparseMatrix
m1 = fromList 5 5 [0,1,11, 3,0,30, 3,1,31]

m2 :: SparseMatrix
m2 = fromList 5 5 [0,1,-11, 3,0,30, 4,1,41]

m3 :: SparseMatrix 
m3 = add m1 m2

m4 :: SparseMatrix
m4 = transpose m3

File: 1cuatri/ED/final/2023-sept/java/sept23/src/sparseMatrix/SparseMatrix.java
/******************************************************************************
 * Student's name: ?????????????????????????????????????
 * Identity number (DNI if Spanish/passport if Erasmus): ???????????????????
 * Student's group: ?
 * PC code: ???
 *
 * Data Structures. Grados en Informatica. UMA.
 *****************************************************************************/

package sparseMatrix;

import dataStructures.dictionary.AVLDictionary;
import dataStructures.dictionary.Dictionary;
import dataStructures.list.List;

    // | = Exercise a - SparseMatrix constructor
public class SparseMatrix {
    public final int rows;
    public final int columns;
    private Dictionary<Index, Integer> nonZeros;

    public SparseMatrix(int r, int c) {
        nonZeros = new AVLDictionary<>();
        rows = r;
        columns = c;
    }

    // | = Exercise b - value
    private int value(Index ind) {
        if(nonZeros.isDefinedAt(ind)){
            return nonZeros.valueOf(ind);
        }
        else{
            return 0;
        }
    }

    // | = Exercise c - update
    private void update(Index ind, int value) {
        nonZeros.insert(ind,value);
    }

    // | = Exercise d - index
    private Index index(int r, int c) {
        if(r>=rows || r<0 || c>=columns || c<0){
            throw new IllegalArgumentException("index: index is not valid");
        }
        return new Index(r,c);
    }

    // | = Exercise e - set
    public void set(int r, int c, int value) {
        Index index = index(r,c);
        this.update(index, value);
    }

    // | = Exercise f - get
    public int get(int r, int c) {
        Index index = index(r,c);
        return this.value(index);
    }

    // | = Exercise g - add
    public static SparseMatrix add(SparseMatrix m1, SparseMatrix m2) {
        SparseMatrix sm = new SparseMatrix(m1.rows, m1.columns);

        for(Index i : m1.nonZeros.keys()){
            sm.set(i.getRow(), i.getColumn(), m1.value(i) + m2.value(i));
        }

        return sm;
    }

    // | = Exercise h - transpose
    public SparseMatrix transpose() {
        SparseMatrix sm = new SparseMatrix(this.rows, this.columns);

        for(Index i : this.nonZeros.keys()){
            sm.set(i.getColumn(), i.getRow(), this.value(i));
            //this.nonZeros.delete(i);
        }

        return sm;
    }

    // | = Exercise i - toString
    public String toString() {
        StringBuilder sb = new StringBuilder();

        for(int i = 0; i<rows; i++){
            for(int j = 0; j<columns; j++){
//                sb.append();

//                while(sb.toString().length() % 6 != 0){
//                    sb.append(" ");
//                }
                sb.append(String.format("%6d",this.value(index(i,j))));

            }
            sb.append('\n');
        }

        return sb.toString();
    }

    // | = Exercise j - fromList and fromList2
    // Complexity using get and ArrayList:
    // Complexity using get and LinkedList:
    public static SparseMatrix fromList(int r, int c, List<Integer> list) {
        if(list.size() % 3 != 0){
            throw new IllegalArgumentException("fromList: list is not multiple of 3");
        }
        SparseMatrix sm = new SparseMatrix(r,c);
        return auxFromList(sm,list);
    }

    private static SparseMatrix auxFromList(SparseMatrix sm, List<Integer> l){
        if(l.isEmpty()){
            return sm;
        }
        else{
            int f = l.get(0);
            l.remove(0);
            int c = l.get(0);
            l.remove(0);
            int v = l.get(0);
            l.remove(0);
            sm.set(f,c,v);

            return auxFromList(sm, l);
        }
    }

    // Complexity using iterator and ArrayList:
    // Complexity using iterator and LinkedList:
    public static SparseMatrix fromList2(int r, int c, List<Integer> list) {
        // todo
        return null;
    }
}
